### DISCUSSION TITLE
great work! How to support?

### DISCUSSION THREAD

**Discussion by UweStrohbeck** [xournalpp/xournalpp#3059] [General]
Hi,
been now for 11 years as university lecturer and found Xournal++ as a great tool which exactely serves the needs. Enhancements through lua plugins already made successfully.
Based on a 40 years programming experience I like to contribute to this project. Unfortunately I didn't found any hints to setup the dev environment. Could somebody share some hints with me? - either on Win or on Linux. 
Br
Uwe

**Comment by rolandlo** [comment]
Hi Uwe! 

How to build the dev environment is described in the wiki: [for Linux](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md) and [for Windows](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md). It's a bit easier on Linux than on Windows.
In #2354 there are some Todo's. Maybe you find something you would like to contribute there. Otherwise just look through the list of [issues](https://github.com/xournalpp/xournalpp/issues). 

Roland

**Comment by UweStrohbeck** [comment]
Hi/Gruezi Roland,

thanks for your help.

Meanwhile I have installed the dev environment on Ubuntu, build successfully based on the actual sources, generated doxygen docs to dive into the source code. Furthermore I picked a todo from the list (custom plugin directory), extended the gui and modified the sources. The tests so far were positive.  At the attached picture you can see the performed gui adaptations and the snippet form the source. I made is customer chooseable, whether to use the standard plugin directory or a custom directory instead.

I didn't made any translations so far to other languages for the gui elements. I didn't tested it neither on Windows nor on Mac. 
What are the next steps to contribute this changes?

Kind regards

Uwe

![Screenshot from 2021-04-21 23-36-00](https://user-images.githubusercontent.com/82940851/115625559-81ea3400-a2fc-11eb-8ed5-ff915845e3b5.png)

**Reply by personalizedrefrigerator** [reply]
Hi,
Thank you for your work!

<details><summary>We accept contributions through pull requests!</summary>


First, ![fork](https://user-images.githubusercontent.com/46334387/115642138-4127f600-a2cf-11eb-9c9b-cfb956db7fc2.png) the project, then push your changes to a branch in the forked git repository.
  * On Linux, do something like this (**I haven't tested this yet!** It might be wrong! I'll test it later and remove this parenthetical if it works):
```bash
$ cd to the root directory of the cloned repository.
$ git branch -b customPluginDirectory
$ git remote add self ssh://LINK_FROM_GITHUB_HERE
$ git add *
$ git commit -m "A descriptive description of the changes here"
$ git push --set-upstream self customPluginDirectory
```
  * Replace LINK_FROM_GITHUB_HERE with the clone URL provided by GitHub on the overview page for **your fork**:
![clone URL](https://user-images.githubusercontent.com/46334387/115642867-91538800-a2d0-11eb-8886-f19ef015084c.png)
   * If you don't have an SSH key set up, see [this document](https://docs.github.com/en/github/authenticating-to-github/adding-a-new-ssh-key-to-your-github-account).
       * You can also use the `https://` link GitHub provides, but then you might have to type in your GitHub password every time you want to sync local changes.
    

--------

If you then go to the overview page for your fork, GitHub should show something like this:
![button for submitting a pull request](https://user-images.githubusercontent.com/46334387/115642424-cca18700-a2cf-11eb-841d-80fabea8db17.png)
Click "Compare and Pull Request".

After submitting a pull request (abbreviated PR), we can review your changes and give feedback.

</details>

> I didn't made any translations so far to other languages for the gui elements.

Don't worry about this. We use [Crowdin](https://crowdin.com/project/xournalpp) to translate GUI changes after they're merged in.

If you have any additional questions, please don't hesitate to ask!

**Comment by UweStrohbeck** [comment]
Hi Henry,
thanks for your help.

Generated a own repository as a fork of master.
But the last command fails

<pre>git push --set-upstream self/customPluginFolder
fatal: &apos;self/customPluginFolder&apos; does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
</pre>

Sorry for bothering you with these simple issues, but I've never worked with git before.

Br
Uwe

**Reply by personalizedrefrigerator** [reply]
It's fine (and other future contributors can refer to this thread, as well)! Git can be really annoying, and my instructions had a mistake!
It's `git push --set-upstream self customPluginFolder`! Sorry (I've corrected the mistake above, too)!

I also added some additional steps above, so I suggest checking it again!

<details><summary>Some Git notes:</summary>

 * Calling the remote repository `self` was arbitrary. It could have been called `myfork` (or something similar) instead (in the `git remote add` stage).
 * I'm also assuming you've run `git add *` (or `git add list of changed files here`) and `git commit -m "Some descriptive commit message here"` to tell Git which changes to sync :)
   * A `commit` is a group of (probably) related changes.
   * If you just type `git commit`, Git will open up an editor that helps you write the commit message.
   * `git add *` is discouraged, because it adds **all** changes made, even unintentional ones.
     * There's a `.gitignore` file that specifies what changes should be ignored, but it doesn't have everything (e.g. `vim` autosave files).
     * Instead, it's recommended that you run `git add path/to/file1 path/to/file2` after making changes.
     * `git status` shows what changes will be added to the next `git commit`, and what still hasn't been added.
 * `--set-upstream` can be shortened to `-u`
 * `git log` shows the commit (change list) history!
 * In the future, if you `git clone` your fork directly, you can just run `git push` to update the fork.
     * By this, I mean `git clone git@github.com:your_username/xournalpp.git` instead of `git clone git@github.com:xournalpp/xournalpp.git` to get the source code.
     * In that case, `git remote add upstream git@github.com:xournalpp/xournalpp.git` could be useful. It gives the upstream repository the name `upstream`. This is useful if you want to sync changes made to the upstream branch while your pull request is still open.
         * E.g. say Xournalpp:xournalpp/master was updated, you might want your fork to also have these changes.
     * To sync changes like this, I suggest the following:
```sh
# Gets changes from upstream, but doesn't apply them.
# (replace upstream with origin if you originally cloned from git@github.com:xournalpp/xournalpp.git)
$ git fetch upstream

# Takes commits you've made and applies them on top of the commit list 
# in upstream.
$ git rebase upstream

# Then, to send the updated history back to GitHub,
$ git push --force
# We need the --force here because we're deleting remote commits, splicing in commits from upstream, then
# adding the deleted commits back on top of the new history.

#  While you could run `git merge upstream` (or `origin` :) ), that
# leads to an additional commit in the history (and applies changes from upstream on top
# of your changes).
#  This would make your changes harder to follow when people are reviewing the pull request.
```

  * Similarly, if you want to edit the commit history, `git rebase` can also be used like this:
```sh
# -i for interactive
# upstream can be replaced with origin.
$ git rebase -i upstream
```
  and Git will allow you to choose which commits (between upstream and the current change (called `HEAD`)) to modify.

</details>

Thank you for asking! I really struggled with this when I first started contributing to Xournal++, so hopefully this will be helpful to other new contributors (so do ask questions like this)!

**Comment by UweStrohbeck** [comment]
Henry - thanks alot!
At least on step further.
Nevertheless, the last step - uploading my changes does not work.
<pre><font color="#8AE234"><b>uwe@uwe-ThinkPad-P40-Yoga</b></font>:<font color="#729FCF"><b>~/cdev/xournalpp</b></font>$ git branch
  customPluginFolder
* <font color="#4E9A06">master</font>
<font color="#8AE234"><b>uwe@uwe-ThinkPad-P40-Yoga</b></font>:<font color="#729FCF"><b>~/cdev/xournalpp</b></font>$ git push --set-upstream self customPluginFolder
Username for &apos;https://github.com&apos;: 
Password for &apos;https://uwestrohbeck@github.com&apos;: 
remote: Permission to xournalpp/xournalpp.git denied to UweStrohbeck.
fatal: unable to access &apos;https://github.com/xournalpp/xournalpp.git/&apos;: The requested URL returned error: 403
<font color="#8AE234"><b>uwe@uwe-ThinkPad-P40-Yoga</b></font>:<font color="#729FCF"><b>~/cdev/xournalpp</b></font>$ 
</pre>
I assume my local git still points to the master branch (I made a master clone yesterday)- which yields now in the access violation while trying to upland changes.
How can I persuade git to use my local customPluginFolder branch for uploading the changes to my personal remote git fork at github.com/UweStrohbeck/xournalpp.git
Br
Uwe

**Comment by UweStrohbeck** [comment]
still learning to cope with git
Could it be, because I made the changes to the master branch instead to my customPluginFolder branch, that **git push** no tries to commit the changes to the master branch/fork/repository whatsoever?

<pre><font color="#8AE234"><b>uwe@uwe-ThinkPad-P40-Yoga</b></font>:<font color="#729FCF"><b>~/cdev/xournalpp</b></font>$ git branch -a -v
  customPluginFolder                          1002c131 [skip ci] Update translation template
* <font color="#4E9A06">master                                     </font> 2b500a6d [ahead 1] custom plugin folder implemented
  <font color="#CC0000">remotes/origin/1.0.x                       </font> b3d46bdf Fix debian description (2nd try)
  <font color="#CC0000">remotes/origin/HEAD                        </font> -&gt; origin/master
  <font color="#CC0000">remotes/origin/debian-hotfix               </font> 0c522cbe [1.0.x] hotfix for debian package
  <font color="#CC0000">remotes/origin/gtk4test                    </font> 1c69cc66 Use resource image
  <font color="#CC0000">remotes/origin/l10n_master                 </font> 5d3522a1 New translations French from Crowdin [skip ci]
  <font color="#CC0000">remotes/origin/master                      </font> 1002c131 [skip ci] Update translation template
  <font color="#CC0000">remotes/origin/refactoring_ElementContainer</font> fe0167f1 fixed a minor bug resulting in a variable is always true
  <font color="#CC0000">remotes/origin/rewrite                     </font> 469e6076 add scroll mask
  <font color="#CC0000">remotes/origin/stable                      </font> 7aa66a78 rewrite comments
  <font color="#CC0000">remotes/origin/view-rewrite                </font> f20ccaa4 Simplify touchscreen scroll/draw mode handling
<font color="#8AE234"><b>uwe@uwe-ThinkPad-P40-Yoga</b></font>:<font color="#729FCF"><b>~/cdev/xournalpp</b></font>$ 
</pre>

If so, how can I solve this 'git challenge'?
Br
Uwe

**Comment by Rmano** [comment]
I do not know if it helps, but this is my recommended procedure for `circuitikz`: https://github.com/circuitikz/circuitikz/blob/master/CONTRIBUTING.md  --- you always push just to your branch, and then prepare a PR (Pull Request) for the main repository.

**Comment by UweStrohbeck** [comment]
Buenas noches Romano,
thanks for the hint. It definitely brought some enlightenment and finally I managed to create a pull request.
My major mistake was right at the beginning. I cloned the master and started to code immediately, without generating a branch.
Nevertheless I managed to push my changes in my local master to my fork on github and generate a pull request afterwards.

The Clang Format Applied check failed - the rest passed successfully.

Br Uwe
